Adding fingerprint table/field endpoints and tests

dragonsahead opened the pull request:
Fixes https://github.com/metabase/metabase/issues/64110 and https://github.com/metabase/metabase/issues/18074

Let's see if this stuff passes the tests

dragonsahead committed:
Add refingerprint endpoints and buttons to refingerprint on metadata models

dragonsahead committed:
Add some tests

dragonsahead committed:
Trying to fix the kondo linter error

dragonsahead committed:
Implement Dan's suggestion to reuse logic from fields-to-refingerprint on refingerprint-table

dragonsahead committed:
Fix formatting issue in fingerprint.clj

camsaul commented on src/metabase/sync/analyze/fingerprint.clj:
```diff
@@ -265,3 +266,11 @@
   (let [table (field/table field)]
     (fingerprint-fields! table [field])))
+
+(mu/defn refingerprint-table
```
```suggestion
(mu/defn refingerprint-table!
```

since it has side-effects

camsaul commented on src/metabase/warehouse_schema_rest/api/field.clj:
```diff
@@ -318,6 +318,20 @@
 
+;; TODO (dragonsahead 2026-01-07) add a response schema to this API endpoint
+#_{:clj-kondo/ignore [:metabase/validate-defendpoint-route-uses-kebab-case
+                      :metabase/validate-defendpoint-has-response-schema]}
```
Let's not introduce new endpoints that are missing response schemas and dig our hole deeper, please add one... it's helpful for our customers to know what API endpoints return (especially important here since this endpoint is ultimately for customer usage)

camsaul commented on test/metabase/warehouse_schema_rest/api/field_test.clj:
```diff
@@ -882,3 +882,15 @@
+      (mt/with-temp [:model/Field {field-id :id} {:name "Field Test"}]
+        (with-redefs [quick-task/submit-task! (fn [task] (task))]
+          (is (= {:status "success"}
+                 (mt/user-http-request :crowberto :post 200 (format "field/%d/refingerprint" field-id)))))))
```
this doesn't actually test that the endpoint does anything other than return this response... you could `with-redefs` `sync/refingerprint-field!` and have it toggle set an `atom` to indicate that this function has actually been called... let me know if you need help with that

camsaul reviewed (changes requested):
- It's very important to include response schemas on all new API endpoints
- All new API endpoints should use `kebab-case` routes and query parameters, you shouldn't need the Kondo ignore because it looks like you already are
- Tests should be tweaked so they test something meaningful
- Please make sure names of functions with side effects end in `!`... a good rule of thumb is that if it's calling a function that ends in `!` then it should end in `!` as well

dragonsahead commented:
@copilot resolve the merge conflicts in this pull request

Copilot committed:
Address review feedback: rename functions, add response schemas, improve tests

Agent-Logs-Url: https://github.com/metabase/metabase/sessions/1b6be98e-9234-4aa2-a7de-0d2baff89cef

Co-authored-by: dragonsahead <1711649+dragonsahead@users.noreply.github.com>

dragonsahead commented:
@copilot rebase this PR against master, resolve merge conflicts and add all response schemas to the new endpoints that we added on this PR

Copilot committed:
Merge remote-tracking branch 'origin/master' into add-fingerprinting-endpoints

# Conflicts:
#	frontend/src/metabase/api/table.ts
#	src/metabase/sync/sync.clj
#	test/metabase/warehouse_schema_rest/api/field_test.clj
#	test/metabase/warehouse_schema_rest/api/table_test.clj

Co-authored-by: dragonsahead <1711649+dragonsahead@users.noreply.github.com>

dragonsahead commented:
@copilot+gpt-5.3-codex please fix the following errors:
Reflection warning, oidc_provider/token_endpoint.clj:75:34 - call to method decode on java.util.Base64$Decoder can't be resolved (argument types: java.lang.Object).
Reflection warning, oidc_provider/token_endpoint.clj:75:25 - call to java.lang.String ctor can't be resolved.
{:clojure.main/message
 "Syntax error reading source at (metabase/warehouse_schema_rest/api/field_test.clj:957:117).\nUnmatched delimiter: )\n",
   [mb.hawk.core$find_tests_for_namespace_symbol invokeStatic "core.clj" 115]
   [mb.hawk.core$find_tests_for_namespace_symbol invoke "core.clj" 113]
   [mb.hawk.core$eval35123$fn__35124 invoke "core.clj" 128]
   [clojure.lang.MultiFn invoke "MultiFn.java" 234]
   [mb.hawk.core$eval35086$fn__35088$fn__35095 invoke "core.clj" 78]
   [clojure.core$map$fn__5956 invoke "core.clj" 2772]
   [clojure.lang.LazySeq force "LazySeq.java" 50]
   [clojure.lang.LazySeq realize "LazySeq.java" 89]
   [clojure.lang.LazySeq seq "LazySeq.java" 106]
   [clojure.lang.Cons next "Cons.java" 41]
   [clojure.lang.RT next "RT.java" 733]
   [clojure.core$next__5472 invokeStatic "core.clj" 64]
   [clojure.core$concat$cat__5581$fn__5582 invoke "core.clj" 744]
   [clojure.lang.LazySeq force "LazySeq.java" 50]
   [clojure.lang.LazySeq realize "LazySeq.java" 89]
   [clojure.lang.LazySeq seq "LazySeq.java" 106]
   [clojure.lang.Cons next "Cons.java" 41]
   [clojure.lang.RT countFrom "RT.java" 673]
   [clojure.lang.RT count "RT.java" 663]
   [mb.hawk.partition$validate_partition_options invokeStatic "partition.clj" 128]
   [mb.hawk.partition$validate_partition_options invoke "partition.clj" 123]
   [mb.hawk.partition$partition_tests invokeStatic "partition.clj" 141]
   [mb.hawk.partition$partition_tests invoke "partition.clj" 135]
   [mb.hawk.core$find_tests_with_options invokeStatic "core.clj" 143]
   [mb.hawk.core$find_tests_with_options invoke "core.clj" 135]
   [mb.hawk.core$find_and_run_tests_with_options invokeStatic "core.clj" 248]
   [mb.hawk.core$find_and_run_tests_with_options invoke "core.clj" 242]
   [mb.hawk.core$find_and_run_tests_cli invokeStatic "core.clj" 286]
   [mb.hawk.core$find_and_run_tests_cli invoke "core.clj" 278]
   [metabase.test_runner$find_and_run_tests_cli invokeStatic "test_runner.clj" 142]
   [metabase.test_runner$find_and_run_tests_cli invoke "test_runner.clj" 138]
   [clojure.lang.Var invoke "Var.java" 386]
   [clojure.run.exec$exec invokeStatic "exec.clj" 88]
   [clojure.run.exec$exec invoke "exec.clj" 78]
   [clojure.run.exec$_main$fn__33329 invoke "exec.clj" 227]
   [clojure.run.exec$_main invokeStatic "exec.clj" 223]
   [clojure.run.exec$_main doInvoke "exec.clj" 191]
   [clojure.lang.RestFn applyTo "RestFn.java" 140]
   [clojure.lang.Var applyTo "Var.java" 707]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.main$main_opt invokeStatic "main.clj" 515]
   [clojure.main$main_opt invoke "main.clj" 511]
   [clojure.main$main invokeStatic "main.clj" 665]
   [clojure.main$main doInvoke "main.clj" 617]
   [clojure.lang.RestFn applyTo "RestFn.java" 140]
   [clojure.lang.Var applyTo "Var.java" 707]
   [clojure.main main "main.java" 40]],
  :cause "Unmatched delimiter: )",
  :phase :read-source}}

Syntax error reading source at (metabase/warehouse_schema_rest/api/field_test.clj:957:117).
Unmatched delimiter: )

Copilot committed:
Fix unmatched delimiter in field_test

Agent-Logs-Url: https://github.com/metabase/metabase/sessions/987791d4-d441-4e71-b7fc-6cdee5bdfa8b

Co-authored-by: dragonsahead <1711649+dragonsahead@users.noreply.github.com>

dragonsahead committed:
Add refingerprint endpoints for tables and fields

Add REST API endpoints `POST /api/field/:id/refingerprint` and
`POST /api/table/:id/refingerprint` that manually trigger an update of
the fingerprints for a Field or all Fields in a Table. The work is
submitted via the quick-task executor and runs with admin perms, matching
the behavior of normal sync.

Backend:
- `metabase.sync.analyze.fingerprint/refingerprint-table!` delegates to
  `fingerprint-table!` with `*refingerprint?*` bound, reusing the existing
  OOM-safe, traced, error-handling logic.
- `metabase.sync.sync/refingerprint-table!` wraps it with a connection
  check (mirroring `refingerprint-field!`).
- New defendpoint routes with response schema `[:map {:closed true}
  [:status [:= :success]]]`.
- API tests verify the endpoints return success, actually invoke the sync
  functions (via with-redefs + atom), and 404 on missing ids.

Frontend:
- RTK Query mutations `refingerprintField`/`refingerprintTable`.
- `RefingerprintFieldButton`/`RefingerprintTableButton` (modeled on the
  existing Rescan buttons), wired into FieldValuesModal and
  SyncOptionsModal, with unit tests.

Fixes #64110 and #18074.

npfitz commented on frontend/src/metabase/metadata/components/RefingerprintTableButton/RefingerprintTableButton.tsx:
```diff
@@ -0,0 +1,33 @@
+  tableId: TableId;
+}
+
+export const RefingerprintTableButton = ({ tableId }: Props) => {
```
This is basically the same component as `RefingerprintFieldButton`. It's not specifically a problem, but it feels like a lot of duplication. Using the `<ActionButton>` component in `FieldValuesModal` and `SyncOptionsModal` would probably be less code overall

npfitz reviewed (commented):
I wouldn't block the PR on my comments below, but I do think there is room for improvement. Otherwise, the FE code looks fine 👍
